File: ts/BLACKJACK.ts
    static deck: number[] = [];
    static playerHand: number[] = [];
    static dealerHand: number[] = [];
    static wager: number = 0;
    static playerBalance: number = 1000;  // Initial player balance

    static main(): void {
        let playAgain: boolean = true;

        while (playAgain) {
            console.log("DO YOU WANT INSTRUCTIONS (IF SO, TYPE 'I')?");
            const input = prompt("Enter choice:");

            if (input?.toUpperCase() === "I") {
                this.printInstructions();
            }

            this.resetGame();
            this.playHand();

            const playAgainInput = prompt("Do you want to play again? (Y/N)");
            playAgain = playAgainInput?.toUpperCase() === "Y";
        }
    }

    static printInstructions(): void {
        console.log("Welcome to Blackjack - Las Vegas style.");
        console.log("The dealer hits on 16 or less and stays on 17 or more.");
        console.log("You may double your bet on your first hit and receive exactly one more card.");
        console.log("Instructions: 0 - No hit, 1 - Hit, 2 - Double.");
    }

    static resetGame(): void {
        this.playerHand = [];
        this.dealerHand = [];

        if (this.deck.length < 15) {
            this.deck = Array.from({ length: 52 }, (_, counter) => (counter % 13) + 1); // Values 1-13
            this.shuffleDeck();
        }
    }

    static shuffleDeck(): void {
        for (let counter = 0; counter < this.deck.length; counter) {
            const swapIndex = Math.floor(Math.random() * this.deck.length);
            [this.deck[counter], this.deck[swapIndex]] = [this.deck[swapIndex], this.deck[counter]];
        }
    }

    static playHand(): void {
        console.log(Current balance: $${this.playerBalance}. Place your wager:`);
        const wagerInput = prompt("Enter wager:");
        this.wager = parseInt(wagerInput || "0");
        this.dealCards();

        if (this.calculateHandValue(this.playerHand) === 21) {
            console.log("Blackjack! You win!");
            this.playerBalance += this.wager * 1.5;
            return;
        } else {
            this.playerTurn();
            if (this.calculateHandValue(this.playerHand) <= 21) {
                this.dealerTurn();
            }
            this.determineWinner();
        }
    }

    static dealCards(): void {
        this.playerHand.push(this.drawCard(), this.drawCard());
        this.dealerHand.push(this.drawCard(), this.drawCard());

        console.log("Your cards:", this.cardString(this.playerHand[0]), "and", this.cardString(this.playerHand[1]));
        console.log("Dealer shows:", this.cardString(this.dealerHand[0]));
    }

    static drawCard(): number {
        return this.deck.pop() || 0; // Safety check if deck is empty
    }

    static cardString(cardValue: number): string {
        const cardNames = ["Ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King"];
        return cardNames[(cardValue - 1) % 13];
    }

    static calculateHandValue(hand: number[]): number {
        let value = 0;
        let aces = 0;

        for (const card of hand) {
            if (card > 10) {
                value += 10; // Face cards are worth 10
            } else if (card === 1) {
                aces++;
                value += 11; // Initially count Ace as 11
            } else {
                value += card;
            }
        }

        while (value > 21 && aces > 0) {
            value -= 10; // Convert Ace from 11 to 1
            aces--;
        }

        return value;
    }

    static playerTurn(): void {
        while (true) {
            console.log("Your hand:", this.playerHand.map(this.cardString).join(", "));
            console.log("Your hand value:", this.calculateHandValue(this.playerHand));

            const choice = prompt("Enter 0 to Stay, 1 to Hit, or 2 to Double:");
            if (choice === "0") {
                console.log("You chose to stay.");
                break;
            } else if (choice === "1") {
                this.playerHand.push(this.drawCard());
                console.log("You hit and drew a card.");
                if (this.calculateHandValue(this.playerHand) > 21) {
                    console.log("You busted!");
                    break;
                }
            } else if (choice === "2") {
                console.log("You chose to double down.");
                this.wager *= 2;
                this.playerHand.push(this.drawCard());
                console.log("Your hand:", this.playerHand.map(this.cardString).join(", "));
                break;
            } else {
                console.log("Invalid input. Please try again.");
            }
        }
    }

    static dealerTurn(): void {
        console.log("Dealer's turn. Dealer reveals:", this.cardString(this.dealerHand[1]));
        while (this.calculateHandValue(this.dealerHand) < 17) {
            this.dealerHand.push(this.drawCard());
            console.log("Dealer hits and draws a card.");
        }

        console.log("Dealer's hand:", this.dealerHand.map(this.cardString).join(", "));
        console.log("Dealer's hand value:", this.calculateHandValue(this.dealerHand));
    }

    static determineWinner(): void {
        const playerValue = this.calculateHandValue(this.playerHand);
        const dealerValue = this.calculateHandValue(this.dealerHand);

        if (playerValue > 21) {
            console.log("You busted! Dealer wins.");
            this.playerBalance -= this.wager;
        } else if (dealerValue > 21 || playerValue > dealerValue) {
            console.log("You win!");
            this.playerBalance += this.wager;
        } else if (playerValue < dealerValue) {
            console.log("Dealer wins.");
            this.playerBalance -= this.wager;
        } else {
            console.log("It's a tie!");
        }
    }
`

File: Java/BLACKJACK.java

/*
 * Blackjack Game - Java Version
 * 
 * This is a simple command-line implementation of Blackjack (Las Vegas style).
 * The player can place a wager, and the game proceeds with standard Blackjack rules.
 * The dealer hits on 16 or less and stands on 17 or more.
 * Players can split a pair, double their bet, or choose to hit or stand.
 * The player wins if they get a higher total than the dealer without busting.
 *
 * Author: Mikael Amare
 * Date: 2024-10-22
 * Version: 1.0
*/

import java.util.Random;
import java.util.Scanner;

public class BLACKJACK {
    static int[] deck = new int[52];
    static int[] playerHand = new int[5];
    static int[] dealerHand = new int[5];
    static int wager;
    static Random random = new Random();
    static Scanner scanner = new Scanner(System.in);
    
    static int playerBalance = 1000; // Starting balance

    public static void main(String[] args) {
        boolean playAgain = true;

        while (playAgain && playerBalance > 0) {
            System.out.println("Your current balance is: $" + playerBalance);
            System.out.println("DO YOU WANT INSTRUCTIONS (IF SO, TYPE 'I')?");
            String input = scanner.nextLine();

            if (input.equalsIgnoreCase("I")) {
                printInstructions();
            }

            resetGame();
            playHand();

            if (playerBalance <= 0) {
                System.out.println("You're out of money! Game over.");
                break;
            }

            System.out.println("Do you want to play again? (Y/N)");
            playAgain = scanner.nextLine().equalsIgnoreCase("Y");
        }
    }
    
    public static void printInstructions() {
        System.out.println("This is a game of Blackjack. Las Vegas style.");
        System.out.println("The dealer must hit on 16 or less and stay on 17 or more.");
        System.out.println("You may split two cards if they are the same and play one hand with each.");
        System.out.println("You may double your bet and receive exactly one more card any time on your first hit.");
        System.out.println("Typing instructions are: 0 - No hit, 1 - Hit, 2 - Double, 3 - Split a pair.");
    }
    
    public static void resetGame() {
        int numDecks = 6; // Standard for Vegas
        deck = new int[52 * numDecks];
        for (int counter = 0; counter < deck.length; counter++) {
            deck[counter] = counter % 13 + 1;
        }
        shuffleDeck();
    }
    
    public static void shuffleDeck() {
        for (int counter = 0; counter < deck.length; counter++) {
            int swapIndex = random.nextInt(deck.length);
            int temp = deck[counter];
            deck[counter] = deck[swapIndex];
            deck[swapIndex] = temp;
        }
    }
        
    public static void dealCards() {
        playerHand[0] = drawCard();
        playerHand[1] = drawCard();
        dealerHand[0] = drawCard();
        dealerHand[1] = drawCard();

        System.out.println("Your first card: " + cardString(playerHand[0]));
        System.out.println("Your second card: " + cardString(playerHand[1]));
        System.out.println("Dealer shows: " + cardString(dealerHand[0]));
    }
    

    public static int drawCard() {
        int cardIndex = random.nextInt(deck.length);
        int card = deck[cardIndex];
        deck[cardIndex] = 0;  // Mark the card as used
        return card;
    }
    
    public static String cardString(int card) {
        switch (card) {
            case 1: return "Ace";
            case 11: return "Jack";
            case 12: return "Queen";
            case 13: return "King";
            default: return String.valueOf(card);
        }
    }
    
    public static int playerTotal() {
        return cardValue(playerHand[0]) + cardValue(playerHand[1]);
    }
    
    public static int cardValue(int card) {
        if (card > 10) return 10;  // Jack, Queen, King are all worth 10
        return card;
    }
    
    public static void playHand() {
        boolean validWager = false;

        // Place the wager
        while (!validWager) {
            try {
                System.out.println("\nPlace your wager:");
                wager = scanner.nextInt();
                scanner.nextLine(); // Consume newline

                if (wager > playerBalance) {
                    System.out.println("\nYou cannot wager more than your balance! Current balance: $" + playerBalance);
                } else if (wager <= 0) {
                    System.out.println("\nWager must be greater than zero!");
                } else if (wager > 500) {
                    System.out.println("\nThat's too much - House limit is $500.");
                } else {
                    validWager = true; // Wager is valid
                }
            } catch (Exception e) {
                System.out.println("\nInvalid input! Please enter a valid wager as a whole number.");
                scanner.nextLine(); // Clear the invalid input
            }
        }

        // Deal cards and check for Blackjack
        System.out.println("\nDealing cards...");
        dealCards();
        if (playerTotal() == 21 && playerHand[2] == 0) {
            System.out.println("\nBlackjack! You win!");
            playerBalance += wager * 1.5;
            return;
        }

        // Player's turn
        playerTurn();

        // Dealer's turn
        dealerTurn();

        // Determine winner
        determineWinner();
    }

    public static void playerTurn() {
        boolean playerBusted = false;
        int total = playerTotal();
        
        while (total < 21) {
            System.out.println("Your total is " + total + ". Hit (1), Double (2), Split (3) Stand (0)?");
            int action = scanner.nextInt();
            scanner.nextLine();  // consume newline
            
            if (action == 1) {
                int newCard = drawCard();
                System.out.println("You drew: " + cardString(newCard));
                total += cardValue(newCard);
            } else {
                break;
            }
            
            if (action == 2) {
                wager *= 2;
                int newCard = drawCard();
                System.out.println("You drew: " + cardString(newCard));
                total += cardValue(newCard);
                break; // Ends the turn after doubling down
            }

            if (action == 3) {
                splitPair();
            }
            

            if (total > 21) {
                System.out.println("You busted with " + total + "!");
                playerBusted = true;
                break;
            }
        }
        
        if (!playerBusted) {
            System.out.println("You stand with " + total + ".");
        }
    }
    
    public static void splitPair() {
        if (playerHand[0] % 13 != playerHand[1] % 13) { // Not a pair
            System.out.println("You cannot split; your cards are not a pair.");
            return;
        }

        int[] splitHand1 = new int[5];
        int[] splitHand2 = new int[5];

        splitHand1[0] = playerHand[0];
        splitHand2[0] = playerHand[1];

        splitHand1[1] = drawCard();
        splitHand2[1] = drawCard();

        System.out.println("Split hand 1: " + cardString(splitHand1[0]) + ", " + cardString(splitHand1[1]));
        System.out.println("Split hand 2: " + cardString(splitHand2[0]) + ", " + cardString(splitHand2[1]));

        System.out.println("Playing hand 1...");
        playIndividualHand(splitHand1);

        System.out.println("Playing hand 2...");
        playIndividualHand(splitHand2);
    }

    public static void playIndividualHand(int[] hand) {
        int total = calculateTotal(hand);
        while (total < 21) {
            System.out.println("Your total is " + total + ". Hit (1) or Stand (0)?");
            int action = scanner.nextInt();
            scanner.nextLine(); // consume newline

            if (action == 1) {
                int newCard = drawCard();
                System.out.println("You drew: " + cardString(newCard));
                hand[findEmptySlot(hand)] = newCard;
                total = calculateTotal(hand);
            } else {
                break;
            }

            if (total > 21) {
                System.out.println("You busted with " + total + "!");
                return;
            }
        }
        System.out.println("You stand with " + total + ".");
    }

    public static int findEmptySlot(int[] hand) {
        for (int counter = 0; counter < hand.length; counter++) {
            if (hand[counter] == 0) {
                return counter;
            }
        }
        return -1; // Shouldn't happen in normal gameplay
    }

    public static int calculateTotal(int[] hand) {
        int total = 0;
        int aceCount = 0;
        for (int card : hand) {
            if (card == 1) {
                aceCount++;
                total += 11;
            } else if (card > 10) {
                total += 10;
            } else {
                total += card;
            }
        }
        while (total > 21 && aceCount > 0) {
            total -= 10;
            aceCount--;
        }
        return total;
    }

    public static void dealerTurn() {
        int total = cardValue(dealerHand[0]) + cardValue(dealerHand[1]);
        System.out.println("Dealer's hidden card: " + cardString(dealerHand[1]));
        
        while (total < 17) {
            int newCard = drawCard();
            System.out.println("Dealer draws: " + cardString(newCard));
            total += cardValue(newCard);
        }
        
        if (total > 21) {
            System.out.println("Dealer busted with " + total + "!");
        } else {
            System.out.println("Dealer stands with " + total + ".");
        }
    }
    
    public static void determineWinner() {
        int playerTotal = calculateTotal(playerHand);
        int dealerTotal = calculateTotal(dealerHand);

        if (playerTotal > 21) {
            System.out.println("You lose.");
            playerBalance -= wager;
        } else if (dealerTotal > 21 || playerTotal > dealerTotal) {
            System.out.println("You win!");
            playerBalance += wager;
        } else if (playerTotal == dealerTotal) {
            System.out.println("It's a tie!");
        } else {
            System.out.println("Dealer wins.");
            playerBalance -= wager;
        }
    }
}
